editor/src/core/workers/parser-printer/parser-printer.ts

```diff
@@ -118,67 +118,90 @@ function buildPropertyCallingFunction(
     default:
       const _exhaustiveCheck: never = attribute
       throw new Error(`Unhandled prop type ${JSON.stringify(attribute)}`)
   }
 }
 
+function jsxAttributeToExpression(attribute: JSXAttribute): TS.Expression {
+  function createExpression(): TS.Expression {
+    switch (attribute.type) {
+      case 'ATTRIBUTE_VALUE':
+        if (typeof attribute.value === 'string') {
+          return TS.createLiteral(attribute.value)
+        } else {
+          return jsonToExpression(attribute.value)
+        }
+      case 'ATTRIBUTE_NESTED_OBJECT':
+        const contents = attribute.content
+        const objectPropertyExpressions: Array<TS.ObjectLiteralElementLike> = contents.map(
+          (prop) => {
+            switch (prop.type) {
+              case 'PROPERTY_ASSIGNMENT':
+                return TS.createPropertyAssignment(
+                  TS.createStringLiteral(prop.key),
+                  jsxAttributeToExpression(prop.value),
+                )
+              case 'SPREAD_ASSIGNMENT':
+                return TS.createSpreadAssignment(jsxAttributeToExpression(prop.value))
+              default:
+                const _exhaustiveCheck: never = prop
+                throw new Error(`Unhandled prop type ${prop}`)
+            }
+          },
+        )
+        return TS.createObjectLiteral(objectPropertyExpressions)
+      case 'ATTRIBUTE_NESTED_ARRAY':
+        const arrayExpressions: Array<TS.Expression> = attribute.content.map((elem) => {
+          switch (elem.type) {
+            case 'ARRAY_SPREAD':
+              return TS.createSpread(jsxAttributeToExpression(elem.value))
+            case 'ARRAY_VALUE':
+              return jsxAttributeToExpression(elem.value)
+            default:
+              const _exhaustiveCheck: never = elem
+              throw new Error(`Unhandled array element type ${elem}`)
+          }
+        })
+        return TS.createArrayLiteral(arrayExpressions)
+      case 'ATTRIBUTE_OTHER_JAVASCRIPT':
+        // SP: This is quite truly the most spectacular fudge I've ever had to create in my entire career.
+        // Creates a representation of the AST that is code which executes the TS factory functions...
+        // ...Then evals it to produce the AST and drills into the result a little.
+        const createExpressionAsString = creator(attribute.javascript)
+        const newExpression = SafeFunction(
+          false,
+          { ts: TS, React: React, addCommentsToNode: addCommentsToNode },
+          `return ${createExpressionAsString}.expression`,
+          [],
+          (e) => {
+            throw e
+          },
+        )()
+        return newExpression
+      case 'ATTRIBUTE_FUNCTION_CALL':
+        return buildPropertyCallingFunction(attribute.functionName, attribute.parameters)
+      default:
+        const _exhaustiveCheck: never = attribute
+        throw new Error(`Unhandled prop type ${JSON.stringify(attribute)}`)
+    }
+  }
+  // Slide the comments onto the expression.
```

Comment:
![](https://media1.tenor.com/images/ec47e44f4d9c8532f0935058b91a1c87/tenor.gif?itemid=10456807)

Reply:
That is how I enter every social situation.